Introduce a chasm entity for worker

rkannan82 opened the pull request:
## What changed?
Added a new chasm component to track lifecycle of a worker and its state.

Key files:
- worker.go: Chasm component.
- statemachine.go: State transition logic.
- tasks.go: Task executors
- proto/v1/message.proto: Worker state definition.

See README.md for details.

Overview:
State: Consists of run-time configuration, activities that are bound to it, etc.

Lifecycle: A worker refers to a worker process and can be in one of 3 states: Active, Inactive, CleanedUp. 
- Active->Active: When a worker is running, it will periodically send a heartbeat to the server to renew its lease and keep it alive.
- Active->Inactive: When the server stops receiving a heartbeat for some time period, it will transition it to Inactive state. It will also request the history server to reschedule activities that were bound to this worker. The rationale is that most likely the worker process has died.
- Inactive->Active: But it is possible that the worker is alive, but due to network partition, it was not able to renew its lease. In this case, when the the worker eventually connects to the server, the server will transition it back into Active state.
- Inactive->CleanedUp: After some pre-defined grace period, the server will transition the worker from inactive to CleanedUp state. This is a terminal state. If the server receives a heartbeat in this state, it will be rejected and treated as an error.

## Why?
To support worker liveness and activity cancellation.

## How did you test it?
- [ ] built
- [ ] run locally and tested manually
- [ ] covered by existing tests
- [x] added new unit test(s)
- [ ] added new functional test(s)

## Potential risks
None. Feature is not yet integrated.

rkannan82 committed:
Skeleton

rkannan82 committed:
Implement heartbeat and timers.

rkannan82 committed:
Register workersession component with chasm runtime.

rkannan82 committed:
Fixes

bergundy commented on chasm/lib/workersession/statemachine.go:
```diff
@@ -0,0 +1,90 @@
+// TransitionHeartbeatReceived handles heartbeat reception, extending the lease and scheduling a new timeout.
+var TransitionHeartbeatReceived = chasm.NewTransition(
+	[]workersessionpb.WorkerSessionStatus{workersessionpb.WORKER_SESSION_STATUS_ACTIVE},
+	workersessionpb.WORKER_SESSION_STATUS_ACTIVE,
```
I wonder if we could transition back from expired to active but what you have now is fine IMHO.

bergundy commented on chasm/lib/workersession/statemachine.go:
```diff
@@ -0,0 +1,90 @@
+
+const (
+	// Time to wait before cleaning up an expired session.
+	expiredSessionCleanupDelay = 10 * time.Minute
```
This should be a dynamic config

bergundy reviewed (commented):
This looks great!

rkannan82 committed:
Address review comments

rkannan82 commented on chasm/lib/workersession/component.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+const (
+	Archetype chasm.Archetype = "WorkerSession"
```
Renamed to Worker. No need to introduce a new terminology.

yiminc commented on chasm/lib/worker/tasks.go:
```diff
@@ -0,0 +1,113 @@
+		return nil // Not inactive, nothing to clean up.
+	}
+
+	e.logger.Info("Cleaning up inactive worker")
```
would this include worker identity info?

rkannan82 replied:
Added a workerid tag and used it for logging.

rkannan82 replied:
Added a workerid tag and used it for logging.

rkannan82 committed:
Handle network partition case by resurrecting the state to active

rkannan82 committed:
Add worker heartbeat proto to state.

rkannan82 committed:
Fix chasm signature change

rkannan82 committed:
Rename component.go to worker.go

rkannan82 committed:
Misc improvements

rkannan82 committed:
Use dynamic config for cleanup delay

rkannan82 committed:
Add RecordHeartbeat to Worker struct.

rkannan82 committed:
Add readme

rkannan82 commented on chasm/lib/worker/statemachine.go:
```diff
@@ -0,0 +1,88 @@
+
+// TransitionHeartbeatReceived handles heartbeat reception, extending the lease and scheduling a new timeout.
+var TransitionHeartbeatReceived = chasm.NewTransition(
+	[]workerpb.WorkerStatus{workerpb.WORKER_STATUS_ACTIVE},
```
After some thought, changing from active to inactive is better. This way SDK does not have to handle it. The only case that SDK needs to handle is inactive -> cleaned up state. The grace period should be large enough to avoid this scenario.

bergundy commented on chasm/lib/worker/library.go:
```diff
@@ -0,0 +1,53 @@
+			l.leaseExpiryTaskExecutor,
+		),
+		chasm.NewRegistrablePureTask(
+			"WorkerCleanupTask",
```
No need to stutter and I see that we are using camel case for the other task names.

```suggestion
			"expiry",
			l.leaseExpiryTaskExecutor,
			l.leaseExpiryTaskExecutor,
		),
		chasm.NewRegistrablePureTask(
			"cleanup",
```

bergundy commented on chasm/lib/worker/statemachine.go:
```diff
@@ -0,0 +1,91 @@
+
+// EventHeartbeatReceived is triggered when a heartbeat is received from the worker.
+type EventHeartbeatReceived struct {
+	Time          time.Time
```
Don't put time on the events, use `ctx.Now(w)` to get the time from the framework. This is critical to allow us to implement time skipping in the framework later.

rkannan82 replied:
I removed time fields from other events also. My intent was for future bookkeeping. But realized that chasm can record this automatically when a transition is created or executed.

bergundy commented on chasm/lib/worker/tasks.go:
```diff
@@ -0,0 +1,121 @@
+	task *workerstatepb.LeaseExpiryTask,
+) error {
+	// Validate that this lease expiry is still relevant.
+	if !e.isLeaseExpiryTaskValid(worker, attrs) {
```
Don't need to validate a pure task, validation should happen under the same lock.

bergundy commented on chasm/lib/worker/library.go:
```diff
@@ -0,0 +1,53 @@
+
+func (l *Library) Components() []*chasm.RegistrableComponent {
+	return []*chasm.RegistrableComponent{
+		chasm.NewRegistrableComponent[*Worker]("Worker"),
```
I see you defined an archetype name and did not use it here.

bergundy commented on chasm/lib/worker/statemachine.go:
```diff
@@ -0,0 +1,91 @@
+
+	taskAttrs := chasm.TaskAttributes{
+		ScheduledTime: leaseDeadline,
+		Destination:   "", // Empty means local execution.
```
nit: this is redundant.

bergundy commented on chasm/lib/worker/tasks.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// isLeaseExpiryTaskValid checks if this lease expiry task is valid or if the lease has been renewed.
+func (e *LeaseExpiryTaskExecutor) isLeaseExpiryTaskValid(
```
nit: this can be a worker method.

bergundy commented on chasm/lib/worker/worker.go:
```diff
@@ -0,0 +1,87 @@
+			LeaseDeadline: leaseDeadline,
+		})
+	case workerstatepb.WORKER_STATUS_INACTIVE:
+		// Handle worker resurrection after network partition
```
This isn't just due to network partition, it can happen when a worker is overloaded, e.g. has large GC freeze. Basically we don't know why...

bergundy commented on chasm/lib/worker/statemachine.go:
```diff
@@ -0,0 +1,91 @@
+// TransitionWorkerResurrection handles worker reconnection when in INACTIVE state.
+// This is a special case for when the same worker process reconnects after network partition.
+// Note: Any activities associated with this worker may have already been rescheduled.
+var TransitionWorkerResurrection = chasm.NewTransition(
```
nit: You can merge this transition with `TransitionActiveHeartbeat` if you want, which may save you some lines of code but what you have is also fine.

rkannan82 replied:
Renamed for now.

bergundy replied:
That's fine, but keep in mind that you are adding redundant status checks in `worker.recordHeartbeat` due to this separation.

bergundy commented on chasm/lib/worker/worker.go:
```diff
@@ -0,0 +1,87 @@
+	if w.WorkerHeartbeat == nil {
+		return ""
+	}
+	return w.WorkerHeartbeat.WorkerInstanceKey
```
Use getters for protos:

```suggestion
	return w.GetWorkerHeartbeat().GetWorkerInstanceKey()
```

bergundy commented on chasm/lib/worker/worker.go:
```diff
@@ -0,0 +1,87 @@
+	w.WorkerHeartbeat = heartbeat
+
+	// Calculate lease deadline
+	leaseDeadline := time.Now().Add(leaseDuration)
```
Pass the duration into the state machine transition or use `ctx.Now(w)` instead or `time.Now()`.

bergundy commented on chasm/lib/worker/tasks.go:
```diff
@@ -0,0 +1,121 @@
+	task *workerstatepb.WorkerCleanupTask,
+) (bool, error) {
+	// Only valid if worker is inactive.
+	return worker.Status == workerstatepb.WORKER_STATUS_INACTIVE, nil
```
You're going to want to avoid multiple cleanup tasks from being considered valid. They could be if you're resurrecting a worker. The simplest way of achieving this is to store the time the statemachine considers itself inactive and compare that to the task attributes schedule time.

bergundy commented on chasm/lib/worker/tasks.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+// Execute is called to clean up inactive workers.
+func (e *WorkerCleanupTaskExecutor) Execute(
```
This will not be a pure task when you have to make RPCs into all of the activity state machines to notify them of completion.

rkannan82 replied:
WorkerCleanupTask just transitions the state to cleaned_up (terminal state). This is my thinking. When lease expires, the LeaseExpiryTaskExecutor will do 2 things:
1. Add RescheduleActivityTask: This will not be a pure task as it notifies all activities.
2. Add WorkerCleanupTask: This is a pure task that just changes the state to cleaned_up.

What do you think?

bergundy replied:
The transition to inactive will probably already need to notify all of the activity state machines that the worker is inactive. No need to create extra tasks if you already have an executor. You want to minimize persistence writes if possible. We will get to that in a follow up PR though. Just keep in mind you are going to have some refactoring here.

rkannan82 committed:
Address review comments.

- Removed time from all events that are passed to transition. It was meant for debugging purposes. Technically, chasm can record this when it invokes the transition function.
- Moved dynamic config into the chasm config file.
- Removed redundant task validation for pure task.

rkannan82 committed:
Reset cleanup grace period when switching from inactive to active.

rkannan82 committed:
Review comments

- Method name is now package internal.

rkannan82 commented on chasm/lib/worker/tasks.go:
```diff
@@ -0,0 +1,121 @@
+		return false
+	}
+
+	if worker.LeaseExpirationTime == nil {
```
It will be a bug if this happens. I added an error log. Should I panic instead of returning false?

rkannan82 committed:
Delete irrelevant file

rkannan82 commented on chasm/lib/worker/README.md:
````diff
@@ -0,0 +1,126 @@
+### States
+
+```
+ACTIVE ──────► INACTIVE ──────► CLEANED_UP
````
Ack.

rkannan82 committed:
Formatting changes

rkannan82 committed:
Rename dynamic config

rkannan82 committed:
Rename tasks.go to executors.go

bergundy commented on chasm/lib/worker/config.go:
```diff
@@ -0,0 +1,27 @@
+	InactiveWorkerCleanupDelay = dynamicconfig.NewNamespaceDurationSetting(
+		"worker.inactiveWorkerCleanupDelay",
+		60*time.Minute,
+		`InactiveWorkerCleanupDelay is the time to wait before cleaning up inactive workers after their lease expires.`,
```
nit:

```suggestion
		"worker.inactiveCleanupDelay",
		60*time.Minute,
		`Time to wait before cleaning up inactive workers after their lease expires.`,
```

rkannan82 committed:
Make inactive as terminal state

rkannan82 committed:
Review comments and cleanup

rkannan82 committed:
Documentation

rkannan82 committed:
Fix Transition.Apply argument order: (worker, ctx, event)

rkannan82 committed:
Add request_response.proto with WorkerService and RecordHeartbeat RPC
